src/librustc_mir/interpret/cast.rs

```diff
@@ -128,95 +134,97 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
+                // Thin or fat pointer that just hast the ptr kind of target type changed.
+                return Ok(*src);
+            } else {
+                // Casting the metadata away from a fat ptr.
+                assert_eq!(src.layout.size, 2 * self.memory.pointer_size());
+                assert_eq!(dest_layout.size, self.memory.pointer_size());
+                assert!(src.layout.ty.is_unsafe_ptr());
+                return match *src {
+                    Immediate::ScalarPair(data, _) => Ok(data.into()),
+                    Immediate::Scalar(..) => bug!(
+                        "{:?} input to a fat-to-thin cast ({:?} -> {:?})",
+                        *src,
+                        src.layout.ty,
+                        cast_ty
+                    ),
+                };
+            }
         }
 
+        // # The remaining source values are scalar.
+
         // For all remaining casts, we either
         // (a) cast a raw ptr to usize, or
         // (b) cast from an integer-like (including bool, char, enums).
         // In both cases we want the bits.
         let bits = self.force_bits(src.to_scalar()?, src.layout.size)?;
-        Ok(self.cast_from_int_like(bits, src.layout, dest_layout)?.into())
+        Ok(self.cast_from_scalar(bits, src.layout, cast_ty).into())
     }
 
-    fn cast_from_int_like(
+    pub(super) fn cast_from_scalar(
         &self,
-        v: u128, // raw bits
+        v: u128, // raw bits (there is no ScalarTy so we separate data+layout)
         src_layout: TyAndLayout<'tcx>,
-        dest_layout: TyAndLayout<'tcx>,
-    ) -> InterpResult<'tcx, Scalar<M::PointerTag>> {
+        cast_ty: Ty<'tcx>,
+    ) -> Scalar<M::PointerTag> {
         // Let's make sure v is sign-extended *if* it has a signed type.
-        let signed = src_layout.abi.is_signed();
+        let signed = src_layout.abi.is_signed(); // also checks that abi is `Scalar`.
         let v = if signed { self.sign_extend(v, src_layout) } else { v };
-        trace!("cast_from_int: {}, {}, {}", v, src_layout.ty, dest_layout.ty);
+        trace!("cast_from_scalar: {}, {} -> {}", v, src_layout.ty, cast_ty);
         use rustc_middle::ty::TyKind::*;
-        match dest_layout.ty.kind {
+        match cast_ty.kind {
             Int(_) | Uint(_) | RawPtr(_) => {
-                let v = self.truncate(v, dest_layout);
-                Ok(Scalar::from_uint(v, dest_layout.size))
+                let size = match cast_ty.kind {
+                    // FIXME: Isn't there a helper for this? The same pattern occurs below.
+                    Int(t) => {
+                        t.bit_width().map(Size::from_bits).unwrap_or_else(|| self.pointer_size())
+                    }
+                    Uint(t) => {
+                        t.bit_width().map(Size::from_bits).unwrap_or_else(|| self.pointer_size())
+                    }
+                    RawPtr(_) => self.pointer_size(),
+                    _ => bug!(),
+                };
```

Comment:
So much work to avoid calling `layout_of` :grin:

Reply:
I learned the hard way that `layout_of` is expensive...